### PR TITLE
feat: add option to parallelize pdfs

### PR DESCRIPTION
**Pull request by hubert-rutkowski85**
This is implementation of https://github.com/Unstructured-IO/unstructured-python-client/issues/17

It adds a boolean `split_pdf_page` to PartitionParameters, which if True, causes the pdf to be split at client side to 1-page chunks, and send to API. The returned elements are joined to a single result list.

In order to run new tests that have been added for it, run `make test-integration-docker`

### CODE REVIEW

File: README.md
Comment (hubert-rutkowski85) [inline README.md:124]: I recognize the bolded part is major downside of this code. It's because it's difficult to recognize if file is pdf, as AFAIK it's not done currently at client. And I don't want to check extension, as it can be missing, plus a lot of other special cases which are handled by backend. Unless it's agreed to be better solution to at least try with the extension, because it probably is assigned in 95% of cases. Other option is to check for PdfStreamError as in test_decorators.py , and if it raises, then steer the partitioning to non-split version. WDYT?

File: src/unstructured_client/utils/_human_split_pdf.py
Comment (badGarnet) [inline src/unstructured_client/utils/_human_split_pdf.py:44]: can we try with process pool please? multi-threading in python involving models that potentially has their own multi-threading (ocr, layout detection) is dangerous. Additionally this also can trip GIL and cause thread lock.

**Comment by Coniferish** [comment]
@hubert-rutkowski85, Can you provide an explanation in the description of the PR the changes you're making? I will start testing these changes locally, but it'd be best to also provide instructions for reviewers on how to test in the future.

File: src/unstructured_client/utils/_human_split_pdf.py
Comment (badGarnet) [reply src/unstructured_client/utils/_human_split_pdf.py:44]: ok, overreacted a bit; this is "just" call api. but the point about potential thread lock stands

**Comment by hubert-rutkowski85** [comment]
@Coniferish just updated description for how to test it, tomorrow will update add description of changes. In my prev project the issues contained description of what the change was about, and PR was only code (to not duplicate the content). If here are different habits here, I will gladly adjust.

**Comment by Coniferish** [comment]
@hubert-rutkowski85, there's no reference to an open issue in the PR, but it would be good to include some kind of summary similar to the template for PRs in Unstructured (another thing I just added to the list for #43)

File: src/unstructured_client/utils/_human_split_pdf.py
Comment (Coniferish) [inline src/unstructured_client/utils/_human_split_pdf.py:25]: Add handling for if users define `request` using a kwarg instead of a positional arg, similar to the `clean_server_url` decorator:

```
server_url: Optional[str] = cast(Optional[str], kwargs.get("server_url"))

if server_url is None and len(args) > SERVER_URL_ARG_IDX:
    server_url = cast(str, args[SERVER_URL_ARG_IDX])
    url_is_in_kwargs = False
```

File: Makefile
Comment (hubert-rutkowski85) [reply Makefile]: jq was used in first version of scripts, I left it here by accident.

File: _test_unstructured_client/test__decorators.py
Comment (hubert-rutkowski85) [reply _test_unstructured_client/test__decorators.py]: Hmm, if you're using the `make test`, `test-unit` or `test-integration` then only those specific ones will be called. It's done through matching "tags", like `unit` or `integration` in test name. Simple, flexible and really convenient. As for spinning the API in docker, it's already done :) See `make test-integration-docker`

File: src/unstructured_client/utils/_human_split_pdf.py
Comment (pawel-kmiecik) [inline src/unstructured_client/utils/_human_split_pdf.py:59]: Just for future:
If we'll like to report progress in future, it'd be better to use `as_completed` instead of `map` and recreate the order of processing (can use some dict, rather simple).
`map` is not showing the real processing ratio, it's likely to work in "spikes" (because it's preserving the order of input)

File: src/unstructured_client/utils/_human_split_pdf.py
Comment (hubert-rutkowski85) [reply src/unstructured_client/utils/_human_split_pdf.py:44]: Hmm, you mean deadlock could happen here? Why? :) I'm not using locks anywhere, and the ThreadPoolExecutor is safe "inside". Maybe you meant potential race condition that can happen in the call_api function which gets sent few shared params:
- request - but it's deep copied to local var, so it's safe for modification.
- retries - those are not copied, I don't modify them and thought partition function also doesn't do it (why would it?) but better to err on the safe side (will also deep copy)
- self - this is sth that I missed! Fortunately looking inside partition, I don't see there any writes, so it seems safe for now, but again, I could deep copy it and it will be safe.

As for processes, those seems unnecesserily heavy for the task here, which is just sending http request - those kinds of IO things are best kept in threads.

File: src/unstructured_client/utils/_human_split_pdf.py
Comment (hubert-rutkowski85) [reply src/unstructured_client/utils/_human_split_pdf.py:59]: Ok, good point. I think about creating new issue for adding callback, as for sure won't be able to do that in this one (about a week till we know hook solution for Speakeasy). And your suggestion will be implemented there.

File: _test_unstructured_client/test__decorators.py
Comment (hubert-rutkowski85) [reply _test_unstructured_client/test__decorators.py:163]: Ok, you're thinking about unit-test kind of approach. That will be difficult to implement here, as its integration test, which by definition is more generic, and with the sheer amount of code, adding multiple versions will result in copy-paste hell. What's most important with this new feature, is to make sure the output we get from split version is the same as would be from non-split, which the test is checking.

As for no of call threads, every test file has a different length in pages(*), which combined with the varying number of 1/2/5 threads, allows to cover all edge cases.

(*) actually list-item and layout-parser-fast both have 2, so one could be skipped, and instead a 1 pager should be added. TODO

**Comment by hubert-rutkowski85** [comment]
> @hubert-rutkowski85, there's no reference to an open issue in the PR, but it would be good to generally include some kind of summary similar to the template for PRs in Unstructured (another thing I just added to the list for #43)

There is :) on right:
image

I've created the branch from issue page, so it's automatically connected. But I do agree that on github UI this information is weakly presented. Gitlab does this in much better way.

File: README.md
Comment (Coniferish) [reply README.md:124]: I think there can be cases where it would make sense to duplicate the code instead of importing a whole package, but I'm fine either way. Was just an idea for more safely handling incorrect files passed unintentionally.

File: src/unstructured_client/sdkconfiguration.py
Comment (Coniferish) [reply src/unstructured_client/sdkconfiguration.py:4]: I'll let @awalker4 weigh in on this since I'm not confident on best practices with that

File: src/unstructured_client/sdkconfiguration.py
Comment (Coniferish) [reply src/unstructured_client/sdkconfiguration.py:6]: No worries. Didn't realize a lot of this was generated code. Not knowing what's generated and what's added by you adds a bit of overhead for reviewing if I don't know if what I'm looking at are changes by you or speakeasy.

File: _test_unstructured_client/test__decorators.py
Comment (Coniferish) [reply _test_unstructured_client/test__decorators.py]: Sounds good. Didn't realize it was already separated

File: _test_unstructured_client/test__decorators.py
Comment (Coniferish) [reply _test_unstructured_client/test__decorators.py:163]: Sounds good. Maybe add more comments or a docstring explaining what's all being tested if you want to keep it as a single test then

File: README.md
Comment (hubert-rutkowski85) [reply README.md:124]: Yes it's important problem here, as I'm sure we'll get lots of irritated users pointing this out. Sure, I'm also for copying small function or two, however, looking at filetype.py it's big, it's not everything-in-single file, it does have a lot of imports, and it will change often future, so unless it's extracted into separate util library, it's too heavy.  

So IMHO the options are two (and could be combined) 1) looking for PdfStreamError and then disabling splitting 2) checking extension- if it's not .pdf then disable splitting. Should cover 98% of cases. @awalker4 ?

File: openapi.json
Comment (hubert-rutkowski85) [reply openapi.json:1]: Hmm, I thought it's already there :) Also, this isn't a really new file in this repo, it was there before last week.

File: README.md
Comment (hubert-rutkowski85) [reply README.md:139]: Copy-pasted into new section.

File: _test_unstructured_client/test__decorators.py
Comment (hubert-rutkowski85) [reply _test_unstructured_client/test__decorators.py:168]: Done.

File: src/unstructured_client/utils/_human_split_pdf.py
Comment (hubert-rutkowski85) [reply src/unstructured_client/utils/_human_split_pdf.py]: Done.

File: README.md
Comment (hubert-rutkowski85) [reply README.md]: Ok, implemented it.

File: src/unstructured_client/utils/_human_split_pdf.py
Comment (hubert-rutkowski85) [reply src/unstructured_client/utils/_human_split_pdf.py]: Done (the deep-copying of all params).

File: src/unstructured_client/sdkconfiguration.py
Comment (awalker4) [reply src/unstructured_client/sdkconfiguration.py:4]: I think it makes sense to commit the autogenerated code here and reduce churn on the next speakeasy pr

**Review by awalker4 — APPROVED** [review]
LGTM, a few notes but approving in advance.

* I think I'd prefer to see the openapi.json stay in the api repo, otherwise it becomes difficult to know where the source of truth is. As long as we maintain this file outside of FastAPI, I think it makes sense to just add `pdf_page_split` as a param over there. 
* We may need to add deepdiff and pypdf to the `make install`
* The tests didn't run for me at first - the PYTHONPATH needs to be `src`

**Comment by scanny** [comment]
> * I think I'd prefer to see the openapi.json stay in the api repo, otherwise it becomes difficult to know where the source of truth is. As long as we maintain this file outside of FastAPI, I think it makes sense to just add `pdf_page_split` as a param over there.

@awalker4 there is an `openapi.json` in the `unstructured-api` repo and also in the `core-product` repo under `/unstructured-api`. So I think it's already difficult to know where the single source of truth (SSOT) is.

What is our best thinking so far of where the SSOT should be given there will essentially be service-dependent API specs (e.g. paid has `similarity-threshold` parameter and freemium or local don't, I expect many more feature-gaps soon)?

File: src/unstructured_client/sdkconfiguration.py
Comment (awalker4) [reply src/unstructured_client/sdkconfiguration.py:4]: Actually, best way to address merge conflicts is probably to remove any local autogenerated code, and just patch in the decorator in this pr.

**Comment by hubert-rutkowski85** [comment]
>  * We may need to add deepdiff and pypdf to the `make install`

Done.
 
>  * The tests didn't run for me at first - the PYTHONPATH needs to be `src`

Yeah I also had problem with running tests at beginning, but don't remember exactly what was the reason and solution. However, make unit-test works for me with PYTHONPATH=src and .

>  * The last speakeasy pr added hooks, which will be our way to insert logic without having to edit the prs every time. It's probably worth a new pr, but keep in mind that we'll want to move our logic over soon.

Good to know. I won't make such changes in this PR, as I want to close it ASAP.

**Comment by awalker4** [comment]
LGTM!

File: src/unstructured_client/sdkconfiguration.py
Comment (hubert-rutkowski85) [reply src/unstructured_client/sdkconfiguration.py]: Ok, removed auto generated code which was causing conflicts on branch. I believe the rest is fine.

**Pull request merged**
